create a new torch.cuda.memory_usage_in_bytes api

yuguo68 opened the pull request:
Summary:
the current torch.cuda.memory_usage returns the memory utilization, more specifically, percent of time over the past sample period global memory being read/written for Nvidia. 

see more details in https://github.com/pytorch/pytorch/issues/140638

Test Plan: added a new unittest

Differential Revision: D65928031




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

hongxiayang commented on docs/source/cuda.rst:
```diff
@@ -28,6 +28,7 @@ torch.cuda
     is_available
     is_initialized
     memory_usage
+    memory_usage_bytes
```
I think it is better to be called `memory_usage_in_bytes`
```suggestion
    memory_usage_in_bytes
```

hongxiayang commented on test/test_cuda.py:
```diff
@@ -3947,6 +3947,20 @@ def test_nvml_get_handler(self):
         self.assertTrue(0 <= torch.cuda.temperature() <= 150)
 
+    @unittest.skipIf(TEST_PYNVML, "pynvml/amdsmi is not available")
+    def test_memory_usage_bytes(self):
```
```suggestion
    def test_memory_usage_in_bytes(self):
```

hongxiayang commented on torch/cuda/__init__.py:
```diff
@@ -1110,6 +1110,15 @@ def _get_amdsmi_device_index(device: Optional[Union[int, Device]]) -> int:
     return idx_map[idx]
 
 
+def _get_amdsmi_memory_usage_bytes(device: Optional[Union[Device, int]] = None) -> int:
```
```suggestion
def _get_amdsmi_memory_usage_in_bytes(device: Optional[Union[Device, int]] = None) -> int:
```

hongxiayang commented on torch/cuda/__init__.py:
```diff
@@ -1150,6 +1159,24 @@ def _get_amdsmi_clock_rate(device: Optional[Union[Device, int]] = None) -> int:
         return clock_info["clk"]
 
 
+def memory_usage_bytes(device: Optional[Union[Device, int]] = None) -> int:
```
```suggestion
def memory_usage_in_bytes(device: Optional[Union[Device, int]] = None) -> int:
```

hongxiayang commented on torch/cuda/__init__.py:
```diff
@@ -1652,6 +1679,7 @@ def addmm_kernel_impl(*args, **kwargs):
     "memory_stats_as_nested_dict",
     "memory_summary",
     "memory_usage",
+    "memory_usage_bytes",
```
```suggestion
    "memory_usage_in_bytes",
```

hongxiayang reviewed (changes requested):
preferring the name `memory_usage_in_bytes`, as you can see it appears in your comments already.

yuguo68 commented:
update the api name to memory_usage_in_bytes

ngimel commented:
As a bikeshed, I'd propose `device_utilization` that would return all the fields in  [nvmlUtilization_t](https://docs.nvidia.com/deploy/nvml-api/structnvmlUtilization__t.html#structnvmlUtilization__t) as a dict, and `device_memory_info` that would return all the fields in [nvmlMemory_v2_t](https://docs.nvidia.com/deploy/nvml-api/structnvmlMemory__v2__t.html#structnvmlMemory__v2__t)

eqy commented on test/test_cuda.py:
```diff
@@ -3947,6 +3947,20 @@ def test_nvml_get_handler(self):
+        torch.cuda.empty_cache()
+        a = torch.cuda.memory_usage_in_bytes()
+        num_bytes = 256 * 1024**2
+        _ = torch.zeros(num_bytes, dtype=torch.int8, device="cuda")
```
Might want to switch this to `.empty()`, as if this is the first `zeros` call it could incur a kernel load which would increase the apparent device memory usage.

yuguo68 replied:
updated to empty in the new version.

eqy commented on test/test_cuda.py:
```diff
@@ -3947,6 +3947,20 @@ def test_nvml_get_handler(self):
+        torch.cuda.synchronize()
+        b = torch.cuda.memory_usage_in_bytes()
+        mem_bytes = b - a
+        self.assertTrue(mem_bytes > num_bytes // 2, mem_bytes < num_bytes * 8)
```
What's the `//2` and `* 8` for? Are they fudge factors?

yuguo68 replied:
probably due to your above comment on memory overhead, in my testing the mem_bytes can be larger than the tensor size especially for small tensor sizes. Hence I added this range as in this test we are primarily interested in the order of the mem, not exact value.

yuguo68 commented:
> As a bikeshed, I'd propose `device_utilization` that would return all the fields in [nvmlUtilization_t](https://docs.nvidia.com/deploy/nvml-api/structnvmlUtilization__t.html#structnvmlUtilization__t) as a dict, and `device_memory_info` that would return all the fields in [nvmlMemory_v2_t](https://docs.nvidia.com/deploy/nvml-api/structnvmlMemory__v2__t.html#structnvmlMemory__v2__t)

thanks for the suggestion! I take a look at the existing api and find that we already have `mem_get_info` which returns the free and total mem with cudart, https://github.com/pytorch/pytorch/blob/02d0c43c3243f967856798fa8622a8727f9d36a3/torch/cuda/memory.py#L705, we also have a `memory_stats` which returns a dict and have standalone api such as `max_memory_reserved` to return the corresponding values of the dict. Hence I think with `device_memory_info` it is still useful to have a standalone api for the used memory size.

yuguo68 committed:
create a new torch.cuda.memory_usage_bytes api (#140719)

Summary:

the current torch.cuda.memory_usage returns the memory utilization, more specifically, percent of time over the past sample period global memory being read/written for Nvidia. 

see more details in https://github.com/pytorch/pytorch/issues/140638

Test Plan: added a new unittest

Reviewed By: xw285cornell

Differential Revision: D65928031

yuguo68 commented:
@pytorchbot merge

yuguo68 commented:
@pytorchbot merge

yuguo68 commented:
@pytorchbot merge -f "no CI failure, a few jobs were canceled"

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change, but the test is flaky on ROCm' -c nosignal

test_cuda.py::TestCudaMallocAsync::test_memory_usage_in_bytes [GH job link](https://github.com/pytorch/pytorch/actions/runs/11861928983/job/33062081352) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/500ce29e4c1e46e0c16bc0b5281b842fb6c819ef)

ngimel commented:
>we also have a memory_stats which returns a dict and have standalone api such as max_memory_reserved to return the corresponding values of the dict. Hence I think with device_memory_info it is still useful to have a standalone api for the used memory size.

The thing I don't like about `memory"*"` apis (too late to do something about `memory_usage` without bc breaking, although maybe we should) is that it confuses stats reported by cuda caching allocator (most of `memory*` APIs) and stats reported by driver (nvml calls). I would prefer different API names to easily distinguish between those, even if you still want a separate api to report the results of  `nvmlDeviceGetMemoryInfo`

yuguo68 commented:
> > we also have a memory_stats which returns a dict and have standalone api such as max_memory_reserved to return the corresponding values of the dict. Hence I think with device_memory_info it is still useful to have a standalone api for the used memory size.
> 
> The thing I don't like about `memory"*"` apis (too late to do something about `memory_usage` without bc breaking, although maybe we should) is that it confuses stats reported by cuda caching allocator (most of `memory*` APIs) and stats reported by driver (nvml calls). I would prefer different API names to easily distinguish between those, even if you still want a separate api to report the results of `nvmlDeviceGetMemoryInfo`

Yes, I also noticed the cache allocator vs nvml issue when looking at the memory_stats. What name do you suggest for this API? More than happy to make the change.

ngimel commented:
I'd have `memory_*` reserved for caching allocator, and `device_*` for nvml, but open for other suggestions, as long as they are clearly distinguishable.

yuguo68 commented:
> I'd have `memory_*` reserved for caching allocator, and `device_*` for nvml, but open for other suggestions, as long as they are clearly distinguishable.

how about device_memory_used? Compared to device_memory_usage, it is more aligned with the `used` in nvmlMemory struct and I think we can probably drop the in_bytes.

yuguo68 commented:
create a new pr https://github.com/pytorch/pytorch/pull/140870
